File: April 2022/April_26.cpp
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        int n = points.size();
        vector<vector<pair<int, int> > > graph(n, vector<pair<int, int> >());
        for(int i = 0; i < n; i++){
            for(int j = i + 1; j < n; j++){
                int dist = abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1]);
                graph[i].push_back(make_pair(dist, j));
                graph[j].push_back(make_pair(dist, i));
            }
        }
        
        bool vis[n];
        memset(vis, false, sizeof(vis));
        int cost = 0;
        
        priority_queue<pair<int, int>, vector<pair<int, int> >, greater<pair<int, int> > > pq;
        pq.push(make_pair(0, 0));
        while(!pq.empty()){
            pair rem = pq.top(); pq.pop();
            if(vis[rem.second])
                continue;
            vis[rem.second] = true;
            cost += rem.first;
            for(auto &nbr : graph[rem.second])
                if(!vis[nbr.second])
                    pq.push(nbr);
        }
        return cost;
    }
};

File: April 2022/April_26.java
class Solution {
    public int minCostConnectPoints(int[][] points) {
        int n = points.length;
        List<int[]>[] graph = new ArrayList[n];
        for(int i = 0; i < n; i++)
                graph[i] = new ArrayList<>();
        for(int i = 0; i < n; i++){
            for(int j = i + 1; j < n; j++){
                int dist = Math.abs(points[i][0] - points[j][0]) + Math.abs(points[i][1] - points[j][1]);
                graph[i].add(new int[]{j, dist});
                graph[j].add(new int[]{i, dist});
            }
        }
        
        boolean[] vis = new boolean[n];
        int cost = 0;
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        pq.add(new int[]{0, 0});
        while(!pq.isEmpty()){
            int[] rem = pq.remove();
            if(vis[rem[0]])
                continue;
            vis[rem[0]] = true;
            cost += rem[1];
            for(int[] nbr : graph[rem[0]])
                if(!vis[nbr[0]])
                    pq.add(nbr);
        }
        
        return cost;
    }
}
